src/main/java/com/timgroup/statsd/UnixStreamClientChannel.java

```diff
@@ -0,0 +1,127 @@
+package com.timgroup.statsd;
+
+import jnr.unixsocket.UnixSocketAddress;
+import jnr.unixsocket.UnixSocketChannel;
+import jnr.unixsocket.UnixSocketOptions;
+
+import java.io.IOException;
+import java.net.SocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.channels.SocketChannel;
+
+/**
+ * A ClientChannel for Unix domain sockets.
+ */
+public class UnixStreamClientChannel implements ClientChannel {
+    private final UnixSocketAddress address;
+    private final int timeout;
+    private final int bufferSize;
+
+    private SocketChannel delegate;
+    private final ByteBuffer delimiterBuffer = ByteBuffer.allocateDirect(Integer.SIZE / Byte.SIZE).order(ByteOrder.LITTLE_ENDIAN);
+
+    /**
+     * Creates a new NamedPipeClientChannel with the given address.
+     *
+     * @param address Location of named pipe
+     */
+    UnixStreamClientChannel(SocketAddress address, int timeout, int bufferSize) throws IOException {
+        this.delegate = null;
+        this.address = (UnixSocketAddress) address;
+        this.timeout = timeout;
+        this.bufferSize = bufferSize;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return delegate.isConnected();
+    }
+
+    @Override
+    public int write(ByteBuffer src) throws IOException {
+        if (delegate == null || !delegate.isConnected()) {
+            connect();
+        }
+
+        int size = src.remaining();
+        if (size == 0) {
+            return 0;
+        }
+        delimiterBuffer.clear();
+        delimiterBuffer.putInt(size);
+        delimiterBuffer.flip();
+
+        try {
+            if (writeAll(delimiterBuffer, true) > 0) {
```

Comment:
If this call times out and returns zero, this method still returns `size` and the sender counts the write as a success, instead of failure.

Reply:
should be fixed